green/result.py

```diff
@@ -42,29 +44,38 @@ class ProtoTest:
     and can pass between processes.
     """
 
-    def __init__(self, test=None):
-        self.module = ""
-        self.class_name = ""
-        self.method_name = ""
-        self.docstr_part = ""
-        self.subtest_part = ""
-        self.test_time = "0.0"
-        # We need to know that this is a doctest, because doctests are very
-        # different than regular test cases in many ways, so they get special
-        # treatment inside and outside of this class.
-        self.is_doctest = False
+    module: str = ""
+    class_name: str = ""
+    method_name: str = ""
+    docstr_part: str = ""
+    subtest_part: str = ""
+    test_time: str = "0.0"
+    failureException = AssertionError
+    description: str = ""
+
+    # Doctests specific attributes:
+    is_doctest: bool = False
+    filename: str | None = None
+    name: str = ""
```

Comment:
⚠️ Doesn't moving these into the class make each of these into class attributes, instead of instance attributes? We really don't want them shared between instances... 😬

Reply:
They way it works in python is that these are mutable and Python will do a copy on write. This mean that they are shared constant, until written to, then on the first write of each instance they get their own copy. This is actually more efficient but should only be done with immutable objects. Integers and strings are immutable.

Try with the python REPL:
```
>>> class X:
...     x: int = 0
>>> X.x
0

>>> first = X()
>>> first.x
0

>>> second = X()
>>> second.x
0

>>> first.x = 42
>>> second.x = 24
>>> X.x
0

>>> first.x
42

>>> second.x
24
```

Reply:
Phew.